### PR TITLE
Some SharedMatrix hardening

### PR DESCRIPTION
**Pull request by vladsud**
Extracting Matrix changes from https://github.com/microsoft/FluidFramework/pull/19692

These are relatively unrelated changes (could break into multiple PRs if desired):
1. More robust APIs not taking garbage in - it's expensive to debug issues when your code misbehaves.
2. Create & export ISharedMatrix
3. Move protocol related things into ops.ts

### CODE REVIEW

File: packages/dds/matrix/src/matrix.ts [hunk 1]
```diff
@@ -435,8 +425,13 @@ export class SharedMatrix<T = any>
 		currentVector: PermutationVector,
 		oppositeVector: PermutationVector,
 		target: SnapshotPath.rows | SnapshotPath.cols,
-		message: IMergeTreeOp,
+		message: IMergeTreeOp | undefined,
```
Comment (anthony-murphy) [inline packages/dds/matrix/src/matrix.ts:428]: why allow undefined, and then assert it is not undefined. seem better to disallow undefined and force the caller not to pass it

Comment (anthony-murphy) [reply packages/dds/matrix/src/matrix.ts:428]: i think this change makes the code harder to understand, than keeping the previous asserts.

Comment (vladsud) [reply packages/dds/matrix/src/matrix.ts:428]: I made this change before there were any checks whatsoever (and code dealt with `any`) - see this change https://github.com/microsoft/FluidFramework/commit/9579629b8d26ea04b5c274fea8425fb14ccaf5f4#diff-87d20d90ea92ec1d7e133b1cc0e9c22433273248e44635baa75e528525f2332fR456
undefined was passed around.

No strong preference on my part, other maybe - less code in final bundle :)

Comment (anthony-murphy) [reply packages/dds/matrix/src/matrix.ts:428]: i'd prefer it was changed back, to keep the code clear that undefined isn't allowed be submitted. i think its worth the bytes in the bundles

Comment (anthony-murphy) [reply packages/dds/matrix/src/matrix.ts:428]: i just fixed the problem all up, no more undefined on this path now:
https://github.com/microsoft/FluidFramework/pull/20009

**Review by anthony-murphy — APPROVED** [review]

File: packages/dds/matrix/src/matrix.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/dds/matrix/src/matrix.ts:428]: which ever goes in first can just merge with the other, so approving

**Pull request merged**